Mathlib/LinearAlgebra/Matrix/SemiringInverse.lean

```diff
@@ -0,0 +1,276 @@
+  refine sum_bij' (fun p _ ↦ f p) (fun p _ ↦ f p) ?_ ?_ hf hf ?_
+  · intro ⟨x, σ⟩ hp
+    rw [mem_sigma, mem_filter, mem_ofSign] at hp ⊢
+    rw [Perm.mul_apply, sign_mul, hp.2.1, sign_swap h, swap_apply_right]
+    exact ⟨mem_univ (σ i), rfl, rfl⟩
+  · intro ⟨x, σ⟩ hp
+    rw [mem_sigma, mem_filter, mem_ofSign] at hp ⊢
+    rw [Perm.mul_apply, sign_mul, hp.2.1, sign_swap h, swap_apply_right]
+    exact ⟨mem_univ (σ i), rfl, rfl⟩
+  · intro ⟨x, σ⟩ hp
+    rw [mem_sigma, mem_filter, mem_ofSign] at hp
+    have key : ({j}ᶜ : Finset n) = disjUnion ({i} : Finset n) ({i, j} : Finset n)ᶜ (by simp) := by
+      rw [singleton_disjUnion, cons_eq_insert, compl_insert, insert_erase]
+      rwa [mem_compl, mem_singleton]
+    simp_rw [key, prod_disjUnion, prod_singleton, Perm.mul_apply, swap_apply_left, ← mul_assoc]
+    rw [mul_comm (A i x) (A i (σ i)), hp.2.2]
+    refine congr_arg _ (prod_congr rfl fun x hx ↦ ?_)
+    rw [mem_compl, mem_insert, mem_singleton, not_or] at hx
+    rw [swap_apply_of_ne_of_ne hx.1 hx.2]
+
+theorem mul_adjp_add_detp : A * adjp 1 A + detp (-1) A • 1 = A * adjp (-1) A + detp 1 A • 1 := by
+  ext i j
+  rcases eq_or_ne i j with rfl | h <;> simp_rw [add_apply, smul_apply, smul_eq_mul]
+  · simp_rw [mul_adjp_apply_eq, one_apply_eq, mul_one, add_comm]
+  · simp_rw [mul_adjp_apply_ne A i j h, one_apply_ne h, mul_zero]
+
+variable {A B}
+
+theorem isAddUnit_mul (hAB : A * B = 1) (i j k : n) (hij : i ≠ j) : IsAddUnit (A i k * B k j) := by
+  revert k
+  rw [← IsAddUnit.sum_univ_iff, ← mul_apply, hAB, one_apply_ne hij]
+  exact isAddUnit_zero
+
+theorem isAddUnit_detp_mul_detp (hAB : A * B = 1) :
+    IsAddUnit (detp 1 A * detp (-1) B + detp (-1) A * detp 1 B) := by
+  suffices h : ∀ {s t}, s ≠ t → IsAddUnit (detp s A * detp t B) from
+    (h (by decide)).add (h (by decide))
+  intro s t h
+  simp_rw [detp, sum_mul_sum, IsAddUnit.sum_iff]
+  intro σ hσ τ hτ
+  rw [mem_ofSign] at hσ hτ
+  rw [← hσ, ← hτ, ← sign_inv] at h
+  replace h := ne_of_apply_ne sign h
+  rw [Ne, eq_comm, eq_inv_iff_mul_eq_one, eq_comm] at h
+  simp_rw [Equiv.ext_iff, not_forall, Perm.mul_apply, Perm.one_apply] at h
+  obtain ⟨k, hk⟩ := h
+  rw [mul_comm, ← Equiv.prod_comp σ, mul_comm, ← prod_mul_distrib,
+    ← mul_prod_erase univ _ (mem_univ k), ← smul_eq_mul]
+  exact (isAddUnit_mul hAB k (τ (σ k)) (σ k) hk).smul_right _
+
+theorem isAddUnit_detp_smul_mul_adjp (hAB : A * B = 1) :
+    IsAddUnit (detp 1 A • (B * adjp (-1) B) + detp (-1) A • (B * adjp 1 B)) := by
+  suffices h : ∀ {s t}, s ≠ t → IsAddUnit (detp s A • (B * adjp t B)) from
+    (h (by decide)).add (h (by decide))
+  intro s t h
+  rw [isAddUnit_iff]
+  intro i j
+  simp_rw [smul_apply, smul_eq_mul, mul_apply, detp, adjp_apply, mul_sum, sum_mul,
+    IsAddUnit.sum_iff]
+  intro k hk σ hσ τ hτ
+  rw [mem_filter] at hσ
+  rw [mem_ofSign] at hσ hτ
+  rw [← hσ.1, ← hτ, ← sign_inv] at h
+  replace h := ne_of_apply_ne sign h
+  rw [Ne, eq_comm, eq_inv_iff_mul_eq_one] at h
```

Comment:
What does `rw [Ne]` mean?

Reply:
I think probably `ne_eq` is the preferred way to write this, but certainly just a nit.